Everyday C#: Null safety tutorials (#53542)
dotnet/docs: docs/csharp/fundamentals/null-safety/nullable-reference-types.md

- *Attributes* on APIs describe more nuanced contracts, such as "this argument can be `null`, but the return value is null only when the argument is null."

The compiler combines these signals to produce diagnostics. Warnings on a non-nullable variable mean the variable might receive `null`. Warnings on a nullable variable mean the code might *dereference* it without a null check. *Dereference* means to use the value the variable refers to—for example, to call a method on it (`variable.Method()`), read a property (`variable.Property`), or index into it (`variable[0]`). Dereferencing a variable that has a value of `null` throws an exception at run time. Either kind of warning means the code's behavior doesn't match its stated design.

## Nullable context

Projects created from recent .NET templates set `<Nullable>enable</Nullable>` in the project file, so the guidance in this article applies as written. If you're working in an older project, open the `.csproj` and check that the `<PropertyGroup>` contains the following line; add it if it's missing:

```xml
<Nullable>enable</Nullable>

> ```

Reviewer: It would be good to link to the **Nullable migration strategies** article here.

:::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="DesignIntent":::

The annotations drive the implementation of `ToString`. Because `FirstName` and `LastName` are non-nullable, the override uses them directly in an [interpolated string](../../language-reference/tokens/interpolated.md) (the `$"..."` syntax that embeds expressions in `{}` placeholders) with no null check. `MiddleName` is nullable, so the override checks it against `null` first and only includes it when it's present. The compiler enforces the difference: code that passes a maybe-null value where a non-nullable one is expected produces a warning, and a constructor that leaves a non-nullable member uninitialized also produces a warning.

## Null-state analysis

The compiler tracks the *null-state* of every expression. The state is one of two values:

- *not-null* — the expression is known to be not `null`.
- *maybe-null* — the expression might be `null`.

> 

Reviewer: No spaces around em dash

A local variable's null-state is updated as the compiler analyzes your code. Two things change it: **assignments** and **null checks**. After an assignment, the variable's null-state matches the expression on the right-hand side. If the expression is null or nullable, the variable becomes maybe-null. If the expression is a non-null literal, the variable becomes not-null. After a null check, the variable's null-state reflects whichever branch is taken.

:::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="NullStateTracking":::

In the preceding example, the first dereference produces a warning because `message` is *maybe-null*. After the assignment to a non-null literal, the compiler knows `message` is *not-null*, so the second dereference is safe.

> Null-state analysis works across `if` checks, [pattern matching](../../language-reference/operators/patterns.md) (expressions such as `is null` or `is { } value` that test the shape of a value), and control flow that loops or returns early:

Reviewer: I don't think `value` is supposed to be there

```suggestion
Null-state analysis works across `if` checks, [pattern matching](../../language-reference/operators/patterns.md) (expressions such as `is null` or `is { }` that test the shape of a value), and control flow that loops or returns early:
```

*Nullable reference types* are a group of features that minimize the chance your code throws <xref:System.NullReferenceException?displayProperty=nameWithType>. You declare which variables are intended to hold `null` and which aren't, and the compiler warns when those declarations don't match how your code uses them. The runtime behavior of your program is unchanged—nullable reference types are entirely a compile-time feature.

Three building blocks work together:

- *Variable annotations* (`string` vs. `string?`) express which references are intended to allow `null`.
- *Null-state analysis* tracks whether the value of an expression is *not-null* or *maybe-null* at each point in your code.
- *Attributes* on APIs describe more nuanced contracts, such as "this argument can be `null`, but the return value is null only when the argument is null."

> The compiler combines these signals to produce diagnostics. Warnings on a non-nullable variable mean the variable might receive `null`. Warnings on a nullable variable mean the code might *dereference* it without a null check. *Dereference* means to use the value the variable refers to—for example, to call a method on it (`variable.Method()`), read a property (`variable.Property`), or index into it (`variable[0]`). Dereferencing `null` throws an exception at run time. Either kind of warning means the code's behavior doesn't match its stated design.

Reviewer: This could be slightly clearer for the new coder

```suggestion
The compiler combines these signals to produce diagnostics. Warnings on a non-nullable variable mean the variable might receive `null`. Warnings on a nullable variable mean the code might *dereference* it without a null check. *Dereference* means to use the value the variable refers to—for example, to call a method on it (`variable.Method()`), read a property (`variable.Property`), or index into it (`variable[0]`). Dereferencing a variable that has a value of `null` throws an exception at run time. Either kind of warning means the code's behavior doesn't match its stated design.
```

## Nullable context

> For new projects targeting .NET 6 or later, the project template includes the `<Nullable>enable</Nullable>` element by default. To enable the feature in an existing project, add it manually:

Reviewer: I wouldn't say this. .NET ships with a lot of templates and there is no guarantee that the code a user is looking at has this in the project. Further, projects migrated from .NET Framework won't have this enabled.

I would just say that if they're working with a newly created project, it's more than likely that nullable context is enabled in the project file with `<Nullable>enable</Nullable>`. And that if it's missing, it can be added manually...

Author: I rewrote this as a note at the introduction to the article.

:::code language="xml" source="snippets/nullable-reference-types/project-snippet.xml":::

For migration approaches that enable the feature gradually, file by file or warning-only first. For more information on migration, see [Nullable migration strategies](../../advanced-topics/update-applications/nullable-migration-strategies.md).

## Express intent with annotations

> When you enable the feature, every reference type variable is *non-nullable* by default. Append `?` to declare a *nullable* reference type:

Reviewer: Nit:

```suggestion
When the feature is enabled, every reference type variable is *non-nullable* by default. Append `?` to declare a *nullable* reference type:
```

:::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="Annotations":::

The annotation doesn't change the runtime type. `string` and `string?` are both <xref:System.String?displayProperty=nameWithType>. The `?` informs the compiler of your design intent. That intent shapes the warnings the compiler produces:

- A non-nullable variable has a default *null-state* of *not-null*. The compiler warns if you assign a value that might be `null`.
- A nullable variable has a default *null-state* of *maybe-null*. The compiler warns if you dereference the variable without first checking it.

Use the annotation to make required and optional values visible in the type system. The following `Person` type declares `FirstName` and `LastName` as non-nullable—every person must have both—and `MiddleName` as nullable, because not everyone has one:

> :::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="DesignIntent":::

Reviewer: Nit:

This code could be improved to show a second person without a middle name. Further, the output of the ToString should be printed as a comment at the end. That will really drive home the concept of how the null check changed the code path. Although that might not really be the focus of the code... So, total nit 😁

The annotations drive the implementation of `ToString`. Because `FirstName` and `LastName` are non-nullable, the override uses them directly in an [interpolated string](../../language-reference/tokens/interpolated.md) (the `$"..."` syntax that embeds expressions in `{}` placeholders) with no null check. `MiddleName` is nullable, so the override checks it against `null` first and only includes it when it's present. The compiler enforces the difference: code that passes a maybe-null value where a non-nullable one is expected produces a warning, and a constructor that leaves a non-nullable member uninitialized also produces a warning.

## Null-state analysis

The compiler tracks the *null-state* of every expression. The state is one of two values:

- *not-null* — the expression is known to be not `null`.
- *maybe-null* — the expression might be `null`.

> A local variable's null-state can change on any line of code. The rule is simple: after an assignment, the variable takes on the null-state of the expression on the right-hand side; after a check against `null`, the variable takes on the null-state implied by the branch you're in. For example, assigning a non-null literal or checking that a variable isn't `null` makes it *not-null*, while assigning `null`, the result of a method whose return type is nullable, or a variable that's currently *maybe-null* makes it *maybe-null*:

Reviewer: For a "simple" rule, I find this entire paragraph very complex and verbose. I iterated with copilot to simplify the wording. How's this?

```suggestion
A local variable's null-state is updated as the compiler analyzes your code. Two things change it: **assignments** and **null checks**. After an assignment, the variable's null-state matches the expression on the right-hand side—if the expression is null or nullable, the variable becomes maybe-null; if the expression is a non-null literal, the variable becomes not-null. After a null check, the variable's null-state reflects whichever branch is taken.:
```

:::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="NullStateTracking":::

> In the preceding example, the first dereference produces a warning because `message` is *maybe-null*. After the assignment to a non-null literal, the compiler knows `message` is *not-null*, so the second dereference is safe. The third statement warns because `originalMessage` was captured while it was still `null`.

Reviewer: I have a hard time following the explanation of the code here, as well as the comments in the code. I have no idea why `_ = originalMessage` is in there.

Author: This was one of those examples where copilot commented out the important code that generated the warning. I updated the sample quite bit.

Null-state analysis works across `if` checks, [pattern matching](../../language-reference/operators/patterns.md) (expressions such as `is null` or `is { } value` that test the shape of a value), and control flow that loops or returns early:

> :::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="FlowAnalysis":::

Reviewer: Regarding the code:
```csharp
    for (Node? current = start; current is not null; current = current.Parent)
    {
        // Inside the loop, the compiler knows current is not-null.
        Console.WriteLine(current.Name);
    }
```

While it's a valid and good way to handle a lot of boilerplate code, the `for` loop here might be confusing for the reader, especially if they're new. Most likely they're used to seeing something like this in a `foreach` loop, and reserving the `for` loop to iterate with an index. Reading this code will probably cause the user to reread it over and over trying to understand what's going on. I'm sure there is a clearer way to do this, such as a simple `while` loop.

```csharp
Node? current = start;
while (current is not null)
{
    // Inside the loop, the compiler knows current is not-null.
    Console.WriteLine(current.Name);

    current = current.Parent;
}

The analysis doesn't trace into the bodies of methods. If you need a method to communicate null-state to its callers, use [nullable analysis attributes](../../language-reference/attributes/nullable-analysis.md) on its signature.

## Override the warnings with `!`

Sometimes you know more than the compiler. The *null-forgiving operator* `!` declares that an expression is *not-null*, even when the analysis says otherwise:

> :::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="NullForgiving":::

Reviewer: This is a really good example, but I think the code would benefit from some more comments. I've put those inline in the code file.

Use `!` sparingly. Each occurrence is a place the compiler can no longer protect you. Prefer adding a null check, restructuring the code, or annotating the relevant API so the compiler reaches the right conclusion on its own.

## Attributes that describe API contracts

Annotations on a parameter or return type aren't always expressive enough. A method might accept a possibly null argument but guarantee a non-null result. A test method might return `true` only when its argument isn't null. Use the [nullable analysis attributes](../../language-reference/attributes/nullable-analysis.md) to convey these contracts:

:::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="NullAnalysisAttributes":::

The <xref:System.Diagnostics.CodeAnalysis.NotNullWhenAttribute> tells the compiler that when `IsPresent` returns `true`, the argument is *not-null*. Inside the `if` block, the compiler treats `value` as *not-null* with no null-forgiving operator required. As of .NET 5, all .NET runtime APIs are annotated, so the analysis benefits any code that calls them.

## Generics

> Generic type parameters without constraints can be reference types or value types, so `T?` follows different rules depending on the type argument:

Reviewer: I think we should link to a generics overview. This is another feature of .NET that is interpreting things differently. If the user is unfamiliar with generics, they may want to know more before reading further.

- If `T` is a non-nullable reference type, `T?` is the corresponding nullable reference type.
- If `T` is a value type, `T?` is the same value type. (Add the `struct` constraint to get <xref:System.Nullable%601> behavior.)
- If `T` is already nullable, `T?` is the same type.

Constraints refine the rules:

- `class` requires a non-nullable reference type.
- `class?` allows either a nullable or a non-nullable reference type.
- `struct` requires a non-nullable value type. `T?` then means <xref:System.Nullable%601>.
- `notnull` requires a non-nullable reference or value type.

> - A base class constraint such as `where T : BaseType` requires a non-nullable reference type that derives from `BaseType`. Append `?` (`where T : BaseType?`) to allow either a nullable or a non-nullable reference type.

Reviewer: I find this very confusing because it's talking about the generic type passed and also the use of that generic type. Or is it? I'm not sure. It's not differentiating what it's talking about, so it makes it sound like the types used are declared as nullable or not-nullable, because when you use `T` it can be `T` or `T?`. So seeing `T` around is just confusing.

For example, on the first statement:

> - If `T` is a non-nullable reference type, `T?` is the corresponding nullable reference type.

Is it talking about the `T` provided on the declaration? `void Method<T>()`. If so, how could it be a non-nullable reference type? All reference types are nullable, right? A `string` **variable** is always nullable because it's a reference type. It's the enablement of the nullability feature and declaration of the variable that give it a non-nullable characteristic, right?

Author: I rewrote this entire section. Now, it introduces terms, then explains the rules briefly, and closes with a note that you probably don't need to remember all the rules all the time.

The constraints help the compiler reason about how a generic type parameter is used:

:::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="Generics":::

## Known pitfalls

Two patterns can leave a non-nullable reference holding `null` without a warning. Both patterns are limitations of the static analysis, not bugs in your code.

> **Default structs.** You can create a struct with non-nullable reference fields by using `default` or `new()`. This approach leaves the struct's fields uninitialized:

Reviewer: I think this should be an H3 section

:::code language="csharp" source="snippets/nullable-reference-types/Program.cs" id="DefaultStructPitfall":::

> The fields hold `null` at run time, but the compiler doesn't warn. The same pitfall extends to *arrays of structs*: a new array initializes every element to the struct's default value, so each element's non-nullable reference fields start as `null`. If you must use a struct, prefer [required members](../../language-reference/keywords/required.md) (members the caller must initialize through an object initializer) or a parameterized constructor that callers must invoke, and populate every element of any array you create.

Reviewer: This should be a new paragraph

```suggestion
The fields hold `null` at run time, but the compiler doesn't warn.

The same pitfall extends to *arrays of structs*: a new array initializes every element to the struct's default value, so each element's non-nullable reference fields start as `null`. If you must use a struct, prefer [required members](../../language-reference/keywords/required.md) (members the caller must initialize through an object initializer) or a parameterized constructor that callers must invoke, and populate every element of any array you create.
```

Also, I don't know how "required" could work? Does that force the user to create every element of the array at declaration?

> **Arrays of references.** A new array of a non-nullable reference type contains all `null` elements until you assign each one:

Reviewer: I think this should be an H3 section
